lizard_auth_client/tests.py

```diff
@@ -1071,6 +1072,114 @@ def test_organisation_index_view_superuser(self):
         response = views.ManageOrganisationIndex.as_view()(request)
         self.assertEqual(response.status_code, 200)
 
+    def get_new_user_sso_mock(self):
+        return mock.patch(
+            'lizard_auth_client.views._new_user_sso_post',
+            return_value={
+                'success': True,
+                'user': {'username': 'root',
+                         'first_name': 'willie',
+                         'last_name': 'wortel',
+                         'email': 'noreply@example.com',
+                         'is_active': True,
+                         'is_staff': False,
+                         'is_superuser': False}})
+
+    def get_new_user_response(self, post_data):
+        request = self.request_factory.post(
+            reverse('lizard_auth_client.management_users_add',
+                    args=[self.organisation_1.id]))
+        request.session = {}
+        request.user = self.user_1
+        request.user.is_superuser = True
+        request.POST = post_data
+
+        return views.ManageUserAddView.as_view()(
+            request, self.organisation_1.id)
+
+    def test_organisation_add_totally_new_user(self):
+        """A manager should be able to add an user
+           with an unknown email address and username
+        """
+        with self.get_new_user_sso_mock():
+            response = self.get_new_user_response(
+                {'username': 'root_new',
+                 'first_name': 'willie',
+                 'email': 'noreply_new@example.com'})
+
+            # Should return an HTTP redirect to detail page
+            self.assertEqual(response.status_code, 302)
+
+    def test_organisation_add_known_exact_match_user(self):
+        """A manager should be able to add an user
+           with an exactly matching email-address/username combination
+        """
+        with self.get_new_user_sso_mock() as sso_mock:
+            # Inject error raised by _new_user_sso_post
+            sso_mock.side_effect = views.UserNotCreatedError(
+                "User could not be created",
+                response=mock.Mock(
+                    status_code=200,
+                    __getitem__=lambda x, y: {
+                        'username': 'root',
+                        'email': 'noreply@example.com'}))
+
+            response = self.get_new_user_response(
+                {'username': 'root',
+                 'first_name': 'willie',
+                 'email': 'noreply@example.com'})
+
+            # Should return an HTTP redirect to detail page
+            self.assertEqual(response.status_code, 302)
+
+    def test_organisation_add_user_with_existing_email_address(self):
+        """A manager should not be able to add an user with an existing
+           email address
+        """
+        with self.get_new_user_sso_mock() as sso_mock:
```

Comment:
Ik heb geen flauw idee wat er hier nou met die side effects en mocks gebeurd, zeker omdat de eigenlijke test buiten de "with" zit.

OF ik moet om 21:53 niet naar pull requests kijken OF het moet wat simpeler opgezet worden. Gewoon `requests.post()` mocken is ook vaak genoeg, bijvoorbeeld.

Reply:
Om te beginnen sowieso niet om 21:53 pull requests bekijken! :P 

De reden waarom ik de functie heb gemocked ipv. direct requests.post/get(), is omdat dit erboven ook zo gedaan is. Echter ben ik ook wel te porren voor requests.post() te mocken... Wat vind jij belangerijker: zelfde stijl aanhouden of eenvoudigere mock van requests.post()?

Ok, Lars is (ook) voor de eenvoudigere mock, gaat het aanpassen ;)

De sideeffect raised de gespecificeerde error op de functie.

De uiteindelijke test -> assert zit wel binnen de with.